compute_tools/src/extension_server.rs

```diff
@@ -260,20 +242,44 @@ pub async fn download_library_file(
     lib_name: &str,
     remote_storage: &GenericRemoteStorage,
     pgbin: &str,
-    all_available_libraries: &HashMap<String, RemotePath>,
+    all_available_libraries: &HashMap<String, Vec<RemotePath>>,
 ) -> Result<()> {
+    let lib_name = get_library_name(lib_name);
     let local_libdir: PathBuf = Path::new(&get_pg_config("--pkglibdir", pgbin)).into();
-    let lib_name_with_ext = enforce_so_end(lib_name);
-    info!("looking for library {:?}", &lib_name_with_ext);
-    match all_available_libraries.get(&*lib_name_with_ext) {
-        Some(remote_path) => {
-            download_helper(remote_storage, remote_path, None, &local_libdir).await?
+    info!("looking for library {:?}", &lib_name);
+    match all_available_libraries.get(&*lib_name) {
+        Some(remote_paths) => {
+            let mut library_download_tasks = Vec::new();
+            for remote_path in remote_paths {
+                let file_path = local_libdir.join(remote_path.object_name().expect("bad object"));
+                if file_path.exists() {
+                    info!("File {:?} already exists. Skipping download", &file_path);
+                } else {
+                    library_download_tasks.push(download_helper(
+                        remote_storage,
+                        remote_path.clone(),
+                        None,
+                        &local_libdir,
+                    ));
+                }
+            }
+            pass_any_error(join_all(library_download_tasks).await)?;
+        }
+        None => {
+            // minor TODO: this logic seems to be somewhat faulty for .so.3 type files?
```

Comment:
looks like comment is outdated

Reply:
the situation here is the following:
`download_library_file("test_lib")` is called
the library doesn't exist remotely
then, we also search for it locally, superficially looking for the file "test_lib.so"

Now that I think about it, maybe this doesn't make sense. 
I feel like the logic should work more like this:

`LOAD test_lib`
1. Look for `test_lib` locally
2. If not found locally, try to download it from S3
3. Look for `test_lib` locally again
4. Throw `"library not found"` error if the library is still not found. 

In particular, if the library is present locally we shouldn't need to talk to S3 at all. 
What do you think?